ostd/src/util/range_counter.rs

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /// Returns the counter at the index.
+    #[cfg(ktest)]
+    pub(crate) fn get(&self, index: usize) -> usize {
+        self.counters.get(&index).cloned().unwrap_or(0)
+    }
+
+    /// Adds one new count for all counters in the range.
+    ///
+    /// Returns ranges that the counter has updated from zero to one.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the range has a negative size.
+    pub(crate) fn add(&mut self, range: &Range<usize>) -> impl Iterator<Item = Range<usize>> {
+        assert!(range.start <= range.end);
+        let mut updated_ranges = Vec::new();
+        let mut reported_end = range.start;
+
+        for i in range.clone() {
+            let counter = self.counters.entry(i).or_insert(0);
+            if *counter != 0 {
+                if reported_end < i {
+                    updated_ranges.push(reported_end..i);
+                }
+                reported_end = i + 1;
+            }
+            *counter += 1;
+        }
+        if reported_end < range.end {
+            updated_ranges.push(reported_end..range.end);
+        }
+
+        updated_ranges.into_iter()
+    }
+
+    /// Removes one count for all counters in the range.
+    ///
+    /// Returns ranges that the counter has updated from one to zero.
+    ///
+    /// # Panics
+    ///
+    /// Panics if
+    ///  - the range has a negative size, or
+    ///  - the range contains a counter that is already zero.
+    pub(crate) fn remove(&mut self, range: &Range<usize>) -> impl Iterator<Item = Range<usize>> {
+        assert!(range.start <= range.end);
+        let mut updated_ranges = Vec::new();
+        let mut reported_end = range.start;
+
+        for i in range.clone() {
+            let counter = self.counters.get_mut(&i).expect("Removing a zero counter");
+            if *counter > 1 {
+                if reported_end < i {
+                    updated_ranges.push(reported_end..i);
+                }
+                reported_end = i + 1;
+            }
+            *counter -= 1;
+            if *counter == 0 {
+                self.counters.remove(&i);
+            }
+        }
+        if reported_end < range.end {
+            updated_ranges.push(reported_end..range.end);
+        }
+
+        updated_ranges.into_iter()
+    }
+}
+
+#[cfg(ktest)]
+mod test {
+    use alloc::vec;
+
+    use super::*;
+    use crate::prelude::*;
+
+    /// Macro to check counter values for multiple ranges
+    /// Usage: check_counter_values!(counter, [15..20, 1], [20..30, 2], [30..35, 1])
+    macro_rules! check_counter_values {
+        ($counter:expr, $([$range:expr, $expected:expr]),* $(,)?) => {
+            $(
+                for i in $range {
+                    assert_eq!($counter.get(i), $expected,
+                        "Counter at index {} should be {}, but got {}",
+                        i, $expected, $counter.get(i));
+                }
+            )*
+        };
+    }
+
+    #[ktest]
+    fn test_add_remove_range() {
```

Comment:
```suggestion
    fn add_remove_range() {
```

Remove `test_` prefixes from `ktest` methods?